code/controllers/autofire.dm

```diff
@@ -0,0 +1,264 @@
+
+/**
+ * # Autofire Subsystem
+ *
+ * Maintains a timer-like system to handle autofiring. Much of this code is modeled
+ * after or adapted from the runechat subsytem.
+ *
+ * Note that this has the same structure for storing and queueing shooter component as the timer subsystem does
+ * for handling timers: the bucket_list is a list of autofire component, each of which are the head
+ * of a circularly linked list. Any given index in bucket_list could be null, representing an empty bucket.
+ */
+SUBSYSTEM_DEF(automatedfire)
+	name = "Automated fire"
+	flags = SS_TICKER | SS_NO_INIT
+	wait = 1
+	priority = FIRE_PRIORITY_AUTOFIRE
+
+	/// world.time of the first entry in the bucket list, effectively the 'start time' of the current buckets
+	var/head_offset = 0
+	/// Index of the first non-empty bucket
+	var/practical_offset = 1
+	/// world.tick_lag the bucket was designed for
+	var/bucket_resolution = 0
+	/// How many shooter are in the buckets
+	var/bucket_count = 0
+	/// List of buckets, each bucket holds every shooter that has to shoot this byond tick
+	var/list/bucket_list = list()
+
+/datum/controller/subsystem/automatedfire/PreInit()
+	bucket_list.len = BUCKET_LEN
+	head_offset = world.time
+	bucket_resolution = world.tick_lag
+
+/datum/controller/subsystem/automatedfire/stat_entry(msg)
+	..("ActShooters:[bucket_count]")
+
+/datum/controller/subsystem/automatedfire/fire(resumed = FALSE)
+
+	if (MC_TICK_CHECK)
+		return
+
+	// Check for when we need to loop the buckets, this occurs when
+	// the head_offset is approaching BUCKET_LEN ticks in the past
+	if (practical_offset > BUCKET_LEN)
+		head_offset += TICKS2DS(BUCKET_LEN)
+		practical_offset = 1
+		resumed = FALSE
+
+	// Check for when we have to reset buckets, typically from auto-reset
+	if ((length(bucket_list) != BUCKET_LEN) || (world.tick_lag != bucket_resolution))
+		reset_buckets()
+		bucket_list = src.bucket_list
+		resumed = FALSE
+
+	// Store a reference to the 'working' shooter so that we can resume if the MC
+	// has us stop mid-way through processing
+	var/static/datum/component/automatedfire/shooter
+	if (!resumed)
+		shooter = null
+
+	// Iterate through each bucket starting from the practical offset
+	while (practical_offset <= BUCKET_LEN && head_offset + ((practical_offset - 1) * world.tick_lag) <= world.time)
+		var/datum/component/automatedfire/bucket_head = bucket_list[practical_offset]
+		if(!shooter)
+			shooter = bucket_head
+
+		while (shooter)
+
+			if(shooter.process_shot())//If we are still shooting, we reschedule the shooter to the next_fire
+				shooter.schedule_shot()
+
+			if (MC_TICK_CHECK)
+				return
+
+			// Break once we've processed the entire bucket
+			shooter = shooter.next
+			if (shooter == bucket_head)
+				break
+
+		// Empty the bucket
+		bucket_list[practical_offset++] = null
+
+/datum/controller/subsystem/automatedfire/Recover()
+	bucket_list |= SSautomatedfire.bucket_list
+
+/datum/controller/subsystem/automatedfire/proc/reset_buckets()
```

Comment:
Docs

Reply:
Same that above, spooky didn't document this.